src/dir_contents.ml

```diff
@@ -434,23 +447,189 @@ let build_mlds_map (d : Super_context.Dir_with_jbuild.t) ~files =
+        Super_context.Dir_with_jbuild.t option
+    | Standalone of File_tree.Dir.t
+                    * Super_context.Dir_with_jbuild.t
+    | Group_root of File_tree.Dir.t
+                    * Super_context.Dir_with_jbuild.t
+
+  let is_standalone = function
+    | Standalone _ | Empty_standalone _ -> true
+    | _ -> false
+
+  let cache = Hashtbl.create 32
+
+  let analyze_stanzas stanzas =
+    List.fold_left stanzas ~init:(None, false) ~f:(fun acc stanza ->
+      let is_group_root, has_modules_consumers = acc in
+      match stanza with
+      | Include_subdirs (loc, x) ->
+        if Option.is_some is_group_root then
+          Loc.fail loc "The 'include_subdirs' stanza cannot appear \
+                        more than once";
+        (Some x, has_modules_consumers)
+      | Library _ | Executables _ | Tests _ ->
+        (is_group_root, true)
+      | _ -> acc)
+
+  let rec get sctx ~dir =
+    match Hashtbl.find cache dir with
+    | Some t -> t
+    | None ->
+      let t =
+        match
+          Option.bind (Path.drop_build_context dir)
+            ~f:(File_tree.find_dir (Super_context.file_tree sctx))
+        with
+        | None -> Empty_standalone None
+        | Some ft_dir ->
+          let project_root = Path.of_local (File_tree.Dir.project ft_dir).root in
+          match Super_context.stanzas_in sctx ~dir with
+          | None ->
+            if dir = project_root ||
+               is_standalone (get sctx ~dir:(Path.parent_exn dir)) then
+              Empty_standalone (Some ft_dir)
+            else
+              Is_component_of_a_group_but_not_the_root None
+          | Some d ->
+            let is_group_root, has_modules_consumers =
+              analyze_stanzas d.stanzas
+            in
+            if is_group_root = Some True then
+              Group_root (ft_dir, d)
+            else if not has_modules_consumers &&
+                    dir <> project_root &&
+                    not (is_standalone (get sctx ~dir:(Path.parent_exn dir)))
+            then
+              Is_component_of_a_group_but_not_the_root (Some d)
+            else
+              Standalone (ft_dir, d)
+      in
+      Hashtbl.add cache dir t;
+      t
+
+  let get_assuming_parent_is_part_of_group sctx ~dir ft_dir =
+    match Hashtbl.find cache (File_tree.Dir.path ft_dir) with
+    | Some t -> t
+    | None ->
+      let t =
+        match Super_context.stanzas_in sctx ~dir with
+        | None -> Is_component_of_a_group_but_not_the_root None
+        | Some d ->
+          let is_group_root, has_modules_consumers =
+            analyze_stanzas d.stanzas
+          in
+          if is_group_root = Some True then
+            Group_root (ft_dir, d)
+          else if has_modules_consumers then
+            Standalone (ft_dir, d)
+          else
+            Is_component_of_a_group_but_not_the_root (Some d)
+      in
+      Hashtbl.add cache dir t;
+      t
+end
+
+let cache = Hashtbl.create 32
+
+let rec get sctx ~dir =
```

Comment:
Why not use `Hashtbl.find_or_add` for this function (and others like it)

Reply:
I guess we could use it here. It's just unfortunate that we end up systematically creating a closure while it's only useful in the not found case.

For other functions, we modify the cache in the not found case through a recursive call. It's not clear what the behavior of `Hashtbl.find_or_add` is in this case should be, so it seems better to avoid it.